### DISCUSSION TITLE
Quasar 2, Vuex 4, Typescript, ESLint, Getter, example/someAction -> Unsafe member access

### DISCUSSION THREAD

**Discussion by tfotis** [quasarframework/quasar#10019] [General - Components / Directives / etc]
I am having trouble getting the property of a Vuex module.
Actually, I am getting the property, but ESLint complains with message `Unsafe member access ['example/someAction'] on an "any" value`.

**How to reproduce?**
- Create a new quasar 2 project with Vuex, Typescript, ESLint
- Enable the Example module in `src/store/index.ts`
- In boot file or Vue component, try to access the getter with `store.getters['example/someAction']`

You will get the error in IDE and console.
```
Unsafe member access ['example/someAction'] on an `any` value.eslint@typescript-eslint/no-unsafe-member-access).
```
By applying `// eslint-disable-next-line @typescript-eslint/no-unsafe-member-access`, the error goes away.

Should it work like this?

**Comment by metalsadman** [comment]
Yeah for now, ts support for vuex is lacking atm. Vuex5 will have better support, but I don't think you can do those magic anymore. Better check the rfc about it down at vue repo, also take a look into pinia.

**Reply by muscleman** [reply]
so i also followed this quasar store setup specified -f ts for a typescript store.
[vuex-store](https://quasar.dev/quasar-cli/vuex-store)

following throws state is undefined errors, am i missing something?

      <template>
        <div>
            <q-toggle v-model="drawerState" />
        </div>
     </template>
` <script>
     import { defineComponent, computed } from 'vue'
     import { useStore } from 'vuex'
     export default defineComponent {
         setup () {
             const $store = useStore()

             const drawerState = computed({
                 get: () => $store.state.showcase.drawerState,
                 set: val => {
                     $store.commit('showcase/updateDrawerState', val)
                 }
             })
            return {
                drawerState
            }
       }
    }
    </script>
`

**Reply by metalsadman** [reply]
Import the `useStore` the one from `src/store/index.ts`.

**Reply by muscleman** [reply]
ok, changed that piece to point at the src/store/index.ts

i also changed the following to a type
`
    export interface StateInterface {
        example: ExampleStateInterface;
    }
`
that stopped the below issue
Unsafe member access ['example/someAction'] on an `any` value.eslint@typescript-eslint/no-unsafe-member-access)


so index.vue compiles with this

`
        const drawerState = computed({
            get: () => $store.state.example.drawerState,
            set: val => {
                $store.commit('showcase/updateDrawerState', val)
            }
        })
`


however now i get at runtime
`
index.vue?e3db:15 Uncaught (in promise) TypeError: Cannot read property 'drawerState' of undefined
    at get (index.vue?e3db:15)
    at ComputedRefImpl.reactiveEffect [as effect] (reactivity.esm-bundler.js?0a2b:42)
    at ComputedRefImpl.get value [as value] (reactivity.esm-bundler.js?0a2b:864)
    at unref (reactivity.esm-bundler.js?0a2b:778)
    at Object.get (reactivity.esm-bundler.js?0a2b:781)
    at Object.get (runtime-core.esm-bundler.js?9e79:6791)
    at eval (index.vue?1e06:29)
    at renderFnWithContext (runtime-core.esm-bundler.js?9e79:1123)
    at hSlot (render.js?cdf5:4)
    at Proxy.eval (QPage.js?9989:50)
`

**Reply by muscleman** [reply]
do i need to create a boot file or something to instantiate the store?

**Reply by muscleman** [reply]
if i console.log(store.state) i see this

![Screenshot from 2021-07-10 19-02-37](https://user-images.githubusercontent.com/4632339/125179189-9c240780-e1b1-11eb-8d3f-f1443bc91d30.png)

**Comment by muscleman** [comment]
woot woot, fixed it.

`
import { store } from 'quasar/wrappers'
import { InjectionKey } from 'vue'
import {
  createLogger,
  createStore,
  Store as VuexStore,
  useStore as vuexUseStore,
} from 'vuex'

import gateway from './gateway'
import { IConfiguration } from './gateway/models/interfaces/IConfiguration'

export interface StateInterface {
  gateway: IConfiguration
}

declare module '@vue/runtime-core' {
  interface ComponentCustomProperties {
    $store: VuexStore<StateInterface>
  }
}

export const storeKey: InjectionKey<VuexStore<StateInterface>> = Symbol('vuex-key')

export default store(function (/* { ssrContext } */) {
  const Store = createStore<StateInterface>({
    modules: {
      gateway
    },
    plugins: [createLogger()],
    strict: !!process.env.DEBUGGING
  })

  return Store;
})

export function useStore() {
  return vuexUseStore(storeKey)
}
`
2 takeaways.

1. typed my StateInterface.
`
export interface StateInterface {
  gateway: IConfiguration
}`

2. wrong thing in my createStore. i had gatewayModule instead of gateway

`
const Store = createStore<StateInterface>({
    modules: {
      gateway
    },
    plugins: [createLogger()],
    strict: !!process.env.DEBUGGING
  })

`

**Reply by donPuerto** [reply]
Pinia its better alternative
